service/history/queue/timer_queue_processor.go

```diff
@@ -0,0 +1,693 @@
+// copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package queue
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pborman/uuid"
+
+	h "github.com/uber/cadence/.gen/go/history"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/collection"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/xdc"
+	"github.com/uber/cadence/service/history/config"
+	"github.com/uber/cadence/service/history/engine"
+	"github.com/uber/cadence/service/history/execution"
+	"github.com/uber/cadence/service/history/shard"
+	"github.com/uber/cadence/service/history/task"
+	"github.com/uber/cadence/service/worker/archiver"
+)
+
+type (
+	timerQueueProcessor struct {
+		shard         shard.Context
+		historyEngine engine.Engine
+		taskProcessor task.Processor
+
+		config                *config.Config
+		isGlobalDomainEnabled bool
+		currentClusterName    string
+
+		metricsClient metrics.Client
+		logger        log.Logger
+
+		status       int32
+		shutdownChan chan struct{}
+		shutdownWG   sync.WaitGroup
+
+		ackLevel               time.Time
+		taskAllocator          TaskAllocator
+		activeTaskExecutor     task.Executor
+		activeQueueProcessor   *timerQueueProcessorBase
+		standbyQueueProcessors map[string]*timerQueueProcessorBase
+		standbyQueueTimerGates map[string]RemoteTimerGate
+	}
+)
+
+// NewTimerQueueProcessor creates a new timer QueueProcessor
+func NewTimerQueueProcessor(
+	shard shard.Context,
+	historyEngine engine.Engine,
+	taskProcessor task.Processor,
+	executionCache *execution.Cache,
+	archivalClient archiver.Client,
+) Processor {
+	logger := shard.GetLogger().WithTags(tag.ComponentTimerQueue)
+	currentClusterName := shard.GetClusterMetadata().GetCurrentClusterName()
+	taskAllocator := NewTaskAllocator(shard)
+
+	activeTaskExecutor := task.NewTimerActiveTaskExecutor(
+		shard,
+		archivalClient,
+		executionCache,
+		logger,
+		shard.GetMetricsClient(),
+		shard.GetConfig(),
+	)
+
+	activeQueueProcessor := newTimerQueueActiveProcessor(
+		currentClusterName,
+		shard,
+		historyEngine,
+		taskProcessor,
+		taskAllocator,
+		activeTaskExecutor,
+		logger,
+	)
+
+	standbyQueueProcessors := make(map[string]*timerQueueProcessorBase)
+	standbyQueueTimerGates := make(map[string]RemoteTimerGate)
+	rereplicatorLogger := shard.GetLogger().WithTags(tag.ComponentHistoryReplicator)
```

Comment:
Is there a HistoryReReplicator? If not, maybe use the ComponentHistoryResender

Reply:
Updated to use `ComponentHistoryResender`